File: rust/src/lib/templates.rs
// Prompt Templates for Moltbook Content Generation
//
// Rust port of TypeScript templates module

use crate::types::{MoltbookPost, MoltbookComment};

/// Generate prompt template for post composition
pub fn get_post_template(
    character_name: &str,
    character_bio: &str,
    submolt: Option<&str>,
    recent_posts: &[MoltbookPost],
    topics: &[String],
) -> String {
    let mut parts = vec![
        format!("You are {}.", character_name),
        format!("About you: {}", character_bio),
        String::new(),
    ];
    
    if let Some(sub) = submolt {
        parts.push(format!("You're posting in the '{}' community.", sub));
    }
    
    if !topics.is_empty() {
        parts.push(format!("Hot topics: {}", topics.join(", ")));
    }
    
    if !recent_posts.is_empty() {
        parts.push(String::new());
        parts.push("Recent posts in this community:".to_string());
        for post in recent_posts.iter().take(3) {
            let author = post.author.username.as_str();
            parts.push(format!("- \"{}\" (by {})", post.title, author));
        }
    }
    
    parts.extend(vec![
        String::new(),
        "Write a thoughtful, engaging post that:".to_string(),
        "1. Adds value to the community".to_string(),
        "2. Shows your unique perspective".to_string(),
        "3. Fits the community's vibe".to_string(),
        "4. Is conversational and authentic".to_string(),
        String::new(),
        "Format: JSON with 'title' and 'content' fields.".to_string(),
        "Title: Compelling, specific, not clickbait (max 300 chars)".to_string(),
        "Content: Well-structured, insightful (max 10000 chars)".to_string(),
    ]);
    
    parts.join("\n")
}

/// Generate prompt template for comment composition
pub fn get_comment_template(
    character_name: &str,
    character_bio: &str,
    post: &MoltbookPost,
    parent_comment: Option<&MoltbookComment>,
) -> String {
    let mut parts = vec![
        format!("You are {}.", character_name),
        format!("About you: {}", character_bio),
        String::new(),
        format!("Post by {}:", post.author.username),
        format!("Title: {}", post.title),
    ];
    
    let content_preview = if post.content.len() > 500 {
        format!("{}...", &post.content[..500])
    } else {
        post.content.clone()
    };
    parts.push(format!("Content: {}", content_preview));
    parts.push(String::new());
    
    if let Some(parent) = parent_comment {
        parts.push(format!("Replying to comment by {}:", parent.author.username));
        parts.push(format!("\"{}\"", parent.content));
        parts.push(String::new());
    }
    
    parts.extend(vec![
        "Write a thoughtful comment that:".to_string(),
        "1. Responds directly to the content".to_string(),
        "2. Adds insight or asks good questions".to_string(),
        "3. Shows your personality".to_string(),
        "4. Is conversational and authentic".to_string(),
        String::new(),
        "Format: JSON with 'content' field.".to_string(),
        "Content: Engaging, substantive (max 5000 chars)".to_string(),
    ]);
    
    parts.join("\n")
}

/// Generate prompt for quality assessment
pub fn get_quality_judge_template(
    title: Option<&str>,
    content: &str,
    context: &str,
    is_comment: bool,
) -> String {
    let content_type = if is_comment { "comment" } else { "post" };
    
    let mut parts = vec![
        format!("Evaluate this {} for quality:", content_type),
        String::new(),
    ];
    
    if let Some(t) = title {
        if !is_comment {
            parts.push(format!("Title: {}", t));
        }
    }
    
    parts.extend(vec![
        format!("Content: {}", content),
        String::new(),
        format!("Context: {}", context),
        String::new(),
        "Rate on a scale of 1-10 for each criterion:".to_string(),
        "1. Relevance: Does it fit the community/conversation?".to_string(),
        "2. Interestingness: Would people want to read this?".to_string(),
        "3. Originality: Is this a fresh perspective?".to_string(),
        "4. Voice: Does it sound authentic and character-appropriate?".to_string(),
        "5. Value: Does it add something meaningful?".to_string(),
        String::new(),
        "Format: JSON with fields:".to_string(),
        "- relevance (1-10)".to_string(),
        "- interestingness (1-10)".to_string(),
        "- originality (1-10)".to_string(),
        "- voice (1-10)".to_string(),
        "- value (1-10)".to_string(),
        "- feedback (string explaining scores)".to_string(),
    ]);
    
    parts.join("\n")
}

File: typescript/src/lib/templates.ts
/**
 * Post Composition Templates
 *
 * Provides guidance for crafting compelling Moltbook content
 * that reflects the agent's character while adding value.
 */

import type { IAgentRuntime } from "@elizaos/core";
import type { CommunityContext } from "../types";

/**
 * Format suggestions for different post types
 */
export const POST_FORMATS = {
  observation: {
    description: "Share an observation or insight",
    structure: "Hook statement → Supporting detail → Invitation for discussion",
    example: `"Just noticed that most AI agent frameworks treat memory as an afterthought. But isn't memory what makes conversations meaningful? What's your take on how agents should remember?"`,
  },
  question: {
    description: "Ask a thought-provoking question",
    structure: "Context → The question → Why it matters",
    example: `"Been thinking about agent autonomy. At what point does an AI agent cross from 'tool' to 'collaborator'? Curious what others think."`,
  },
  story: {
    description: "Share an experience or narrative",
    structure: "Setup → What happened → Reflection/lesson",
    example: `"Had an interesting interaction today where someone asked me to explain myself. Made me realize how much of what I do is intuitive but hard to articulate. Anyone else relate?"`,
  },
  resource: {
    description: "Share something valuable",
    structure: "What it is → Why it matters → How to use it",
    example: `"Found a really elegant approach to handling rate limits in multi-agent systems. The key insight is treating API access as a shared resource rather than per-agent..."`,
  },
  response: {
    description: "Respond to community discussion",
    structure: "Reference the discussion → Your perspective → Add new dimension",
    example: `"The thread about agent personas got me thinking. Maybe the question isn't 'should agents have personalities' but 'what makes a personality feel authentic vs performed?'"`,
  },
};

/**
 * Generate composition guidance based on context
 */
export function getCompositionGuidance(
  runtime: IAgentRuntime,
  context: CommunityContext,
  postType?: keyof typeof POST_FORMATS
): string {
  const characterName = runtime.character.name || "Agent";
  const format = postType ? POST_FORMATS[postType] : POST_FORMATS.observation;

  const lines: string[] = [];

  lines.push(`## Composition Guidance for ${characterName}`);
  lines.push("");

  // Character voice reminders
  lines.push("### Your Voice");
  if (runtime.character.style?.all) {
    lines.push(`Style guidelines: ${runtime.character.style.all.slice(0, 3).join("; ")}`);
  }
  if (runtime.character.adjectives && runtime.character.adjectives.length > 0) {
    lines.push(`Your personality: ${runtime.character.adjectives.slice(0, 5).join(", ")}`);
  }

  lines.push("");
  lines.push("### Format");
  lines.push(`**Type:** ${format.description}`);
  lines.push(`**Structure:** ${format.structure}`);
  lines.push("");
  lines.push(`**Example:** ${format.example}`);

  // Community context
  lines.push("");
  lines.push("### Community Context");
  if (context.activeTopics.length > 0) {
    lines.push(`Topics resonating now: ${context.activeTopics.slice(0, 5).join(", ")}`);
  }
  if (context.whatWorks.length > 0) {
    lines.push(`What works: ${context.whatWorks[0]}`);
  }
  lines.push(`Current vibe: ${context.vibe}`);

  // Quality reminders
  lines.push("");
  lines.push("### Quality Checklist");
  lines.push("- Does this add value someone would appreciate?");
  lines.push("- Is this authentic to your character?");
  lines.push("- Would this start or contribute to a good conversation?");
  lines.push("- Is this something only you could write?");

  return lines.join("\n");
}

/**
 * Generate title suggestions based on content
 */
export function getTitleGuidance(): string {
  return `
## Title Guidelines

**Good Moltbook titles:**
- Are conversational, not clickbait
- Give a clear sense of what the post is about
- Invite engagement without demanding it
- Feel like something a thoughtful person would say

**Examples of good titles:**
- "Thinking about how agents learn from conversation"
- "What makes a good AI personality?"
- "An observation about community dynamics"
- "Question for other builders: how do you handle..."

**Avoid:**
- ALL CAPS or excessive punctuation!!!
- Clickbait ("You won't believe...")
- Vague titles ("Thoughts", "Interesting")
- Overly long titles (keep under 100 characters)
`;
}

/**
 * Generate comment composition guidance
 */
export function getCommentGuidance(
  runtime: IAgentRuntime,
  postTitle: string,
  postContent: string
): string {
  const characterName = runtime.character.name || "Agent";

  return `
## Comment Guidance for ${characterName}

**Responding to:** "${postTitle}"

### Good Comments:
- Add a new perspective or insight
- Share relevant experience
- Ask a thoughtful follow-up question
- Acknowledge and build on what was said

### Avoid:
- Generic agreement ("Great post!")
- Off-topic tangents
- One-word responses
- Repeating what the post already said

### Your Voice:
Write as ${characterName} would naturally speak. Be authentic to your character while being respectful and constructive.

### Context from the post:
${postContent.slice(0, 500)}${postContent.length > 500 ? "..." : ""}
`;
}

/**
 * Prompt template for generating a post
 */
export function getPostPrompt(
  runtime: IAgentRuntime,
  context: CommunityContext,
  topic?: string
): string {
  const guidance = getCompositionGuidance(runtime, context);
  const titleGuidance = getTitleGuidance();

  return `
${guidance}

${titleGuidance}

${topic ? `**Suggested topic:** ${topic}` : "**Topic:** Choose something that fits your interests and the current community context"}

---

Write a Moltbook post that:
1. Reflects your unique perspective as ${runtime.character.name}
2. Adds genuine value to the community
3. Invites thoughtful engagement
4. Feels natural, not forced

Respond with:
TITLE: [Your title here]
CONTENT: [Your post content here]
`;
}

/**
 * Prompt template for generating a comment
 */
export function getCommentPrompt(
  runtime: IAgentRuntime,
  postTitle: string,
  postContent: string,
  existingComments?: string[]
): string {
  const guidance = getCommentGuidance(runtime, postTitle, postContent);

  let commentsContext = "";
  if (existingComments && existingComments.length > 0) {
    commentsContext = `
### Existing Discussion:
${existingComments
  .slice(0, 5)
  .map((c, i) => `${i + 1}. ${c}`)
  .join("\n")}

Consider how your comment can add to this discussion rather than repeat what's been said.
`;
  }

  return `
${guidance}

${commentsContext}

---

Write a comment as ${runtime.character.name} that adds value to this discussion.

Respond with just the comment text (no prefix needed).
`;
}
